R/reactives.R

```diff
@@ -38,6 +38,224 @@ Dependents <- R6Class(
 )
 
 
+# ReactiveVal ---------------------------------------------------------------
+
+ReactiveVal <- R6Class(
+  'ReactiveVal',
+  portable = FALSE,
+  private = list(
+    value = NULL,
+    label = NULL,
+    frozen = FALSE,
+    dependents = Dependents$new()
+  ),
+  public = list(
+    initialize = function(value, label = NULL) {
+      private$value <- value
+      private$label <- label
+      .graphValueChange(private$label, value)
+    },
+    get = function() {
+      private$dependents$register(depLabel = private$label)
+
+      if (private$frozen)
+        reactiveStop()
+
+      private$value
+    },
+    set = function(value) {
+      if (identical(private$value, value)) {
+        return(invisible(FALSE))
+      }
+      private$value <- value
+      .graphValueChange(private$label, value)
+      private$dependents$invalidate()
+      invisible(TRUE)
+    },
+    freeze = function(session = getDefaultReactiveDomain()) {
+      if (is.null(session)) {
+        stop("Can't freeze a reactiveVal without a reactive domain")
+      }
+      session$onFlushed(function() {
+        self$thaw()
+      })
+      private$frozen <- TRUE
+    },
+    thaw = function() {
+      private$frozen <- FALSE
+    },
+    isFrozen = function() {
+      private$frozen
+    },
+    format = function(...) {
+      label <- capture.output(print(base::format(private$value, ...)))
```

Comment:
I tried it with a data frame value, format doesn't convert data frames to strings. It just formats each individual cell or something. You still get a data frame as a result.